Feature/interface changes

karivatj opened the pull request:
Hello,

I've been working on the refactoring of the `android-gpgs` module and I needed to make some interface changes to the code. This PR contains these proposed changes to the `IGameService` interface and all necessary changes to sub-modules which implement this interface. I've tested these changes locally and seems to work fine and builds ok too.

Here are the key changes of the PR:

1) Add `resultCode` Integer into `gsOnSessionActive` and `gsOnSessionInactive` methods. I found it is important to know the reason code why these session changes occur. For example, considering the new way to handle logins in Google Play Services, one should consider a situation when the user asks to explicitly sign out from the service or cancel the sign in flow. When this happens, one shouldn't try to sign in after that until the user explicitly wants to sign in by pressing a sign in button in the application.

Thus, I figured I'd add a `resultCode` information in to the callback methods that deliver this information to the caller. This way the caller can persist the status and can make a informed decision whether to sign the user in during app startup or during resume. This is described in [Google Play Games Checklist](https://developers.google.com/games/services/checklist#3_leaderboards) ID 1.6

2) Add get player data method to `IGameService` interface and implement method stubs to sub-modules

This is used to fetch data from the Game Service if it supports it. This was a missing feature, and I thought I'd implement it in
`android-gpgs` module. This is purely optional feature but nice to have.

Add IPlayerData and IPlayerDataResponseListener interfaces which define the data model and the callback methods when 
new player data is received from the Game Service.

3) Add timespan and collection parameters to fetchLeaderboardEntries in IGameService

If the Game Service supports it, these integers can control how to refine the leaderboard
results.

- Timespan declares if for example daily, weekly or all time results are fetched.
- Collection defines if the results retrieved are in public or social context.

This feature is at least supported by Google Play Games and should be added to make more refined queries to leaderboards. Now it defaults to All Time scores and completely disables the possibility to fetch Daily or Weekly leaderboards.

4) Add missing Javadoc stuff to get rid of warnings during compile time.

MrStahlfelge commented on android-amazongc/src/de/golfgl/gdxgamesvcs/GameCircleClient.java:
```diff
@@ -100,7 +101,7 @@ protected void agcServiceNotReady(AmazonGamesStatus amazonGamesStatus) {
         isConnected = false;
         if (gsListener != null) {
-            gsListener.gsOnSessionInactive();
+            gsListener.gsOnSessionInactive(null);
```
Why `null`? Wasn't this a successfull operation? Same for all over calls with null.

karivatj replied:
Went through each callback and added correct response codes

MrStahlfelge commented on core/src/de/golfgl/gdxgamesvcs/player/IPlayerData.java:
```diff
@@ -0,0 +1,44 @@
+package de.golfgl.gdxgamesvcs.player;
+
+/**
+ * Leaderboard entry data
```
Not a leaderboard

karivatj replied:
Fixed

MrStahlfelge commented on core/src/de/golfgl/gdxgamesvcs/player/IPlayerData.java:
```diff
@@ -0,0 +1,44 @@
+public interface IPlayerData {
+
+    /**
+     * Player Id
```
It's redundant information, just doc the return value

karivatj replied:
Done

MrStahlfelge commented on core/src/de/golfgl/gdxgamesvcs/player/IPlayerData.java:
```diff
@@ -0,0 +1,44 @@
+     *
+     * @return Player Retrieved Timestamp
+     */
+    public long getRetrievedTimestamp();
```
what is a retrieved timestamp?

karivatj replied:
From the [docs](https://developers.google.com/android/reference/com/google/android/gms/games/Player#getRetrievedTimestamp())

`Retrieves the timestamp at which this player record was last updated locally.`

MrStahlfelge commented on android-gpgs/src/de/golfgl/gdxgamesvcs/GpgsPlayerData.java:
```diff
@@ -0,0 +1,42 @@
+ * Created by Kari Vatjus-Anttila on 27.11.2021
+ */
+
+public class GpgsPlayerData implements IPlayerData {
```
Since this class is basically a generic implementation if IPlayerData, I wouldn't mind to have it in core. That's not done for other classes, but I think it would be better when it had been done that way. :)

MrStahlfelge commented on core/src/de/golfgl/gdxgamesvcs/IGameServiceClient.java:
```diff
@@ -150,27 +161,32 @@
      * @param relatedToPlayer only fetch scores around current player score or by current player (depending on Game
      *                        Service). If this is not possible, player-unrelated leaderboard entries are returned.
-     * @param callback
+     * @param callback        the listener that will be notified about the result
```
To remain backwards compatibility, we should keep a method with the old parameters.

karivatj replied:
Kept both versions of the method for backwards compatibility. If GameService supports timespans and collections the overloaded method can be used. If not, an exception is thrown.

MrStahlfelge commented:
> 1. Add `resultCode` Integer into `gsOnSessionActive` and `gsOnSessionInactive` methods. I found it is important to know the reason code why these session changes occur. For example, considering the new way to handle logins in Google Play Services, one should consider a situation when the user asks to explicitly sign out from the service or cancel the sign in flow. When this happens, one shouldn't try to sign in after that until the user explicitly wants to sign in by pressing a sign in button in the application.

This was handled implicitely by GpgsClient before, the behaviour should be kept. It's the difference between calling `resumeSession` and `logIn`.
If it didn't work this way on one if the implementations, that was a bug. I know for sure it works on Gpgs Android though. :)

karivatj commented on core/src/de/golfgl/gdxgamesvcs/IGameServiceClient.java:
```diff
@@ -96,6 +97,16 @@
+     *
+     * @param callback the listener that will be notified about the result
+     *
+     * @return Returns false if data retrieve failed. True if successful.
```
I tweaked the Javadoc to this: _@return False if the data could not be fetched. True if the event was triggered successfully._

karivatj commented on android-amazongc/src/de/golfgl/gdxgamesvcs/GameCircleClient.java:
```diff
@@ -242,6 +243,13 @@ public String getPlayerDisplayName() {
+    public boolean getPlayerData(IPlayerDataResponseListener callback) {
+        if(callback != null)
+            callback.onPlayerDataResponse(null);
+        return false;
```
Removed the if clause.

karivatj commented on core/src/de/golfgl/gdxgamesvcs/IGameServiceClient.java:
```diff
@@ -150,27 +161,32 @@
      */
     boolean fetchLeaderboardEntries(String leaderBoardId, int limit, boolean relatedToPlayer,
-                                    IFetchLeaderBoardEntriesResponseListener callback);
+                                    IFetchLeaderBoardEntriesResponseListener callback, int timespan, int collection);
```
Defined two enum classes within IGameService LeaderboardTimespan and LeaderboardCollection. Also added timespan and collection into supported features list.

karivatj commented:
Thank you for the comments. I have made some fixes and will update the PR soon. I will look into the GPGS login stuff a bit more to figure out a way to handle explicit sign out situations within the client code. I had some problems dealing with that stuff when updating the code to use the newer API. For example the old connect() method does not exist anymore and is replaced by silentSignIn() which seems to sign in the user no matter what is the situation (even if the user explicitly signed out previously).

karivatj commented on core/src/de/golfgl/gdxgamesvcs/IGameServiceClient.java:
```diff
@@ -96,6 +97,16 @@
      */
     String getPlayerDisplayName();
```
I've flagged the method with @Deprecated. Although one should keep in mind that only `android-gpgs` implements the new `getPlayerData` method currently (the next, and final PR when this gets approved. Hopefully). `getPlayerData` should be implemented in other game clients also.

karivatj commented on core/src/de/golfgl/gdxgamesvcs/IGameServiceListener.java:
```diff
@@ -12,13 +12,17 @@
+     * @param resultCode result code of the session represented as an int
      */
-    public void gsOnSessionActive();
+    public void gsOnSessionActive(Integer resultCode);
```
Noticed that there is a GsErrorType enum already present. I propose we rename this to GsResultCode which contains result codes for successes also. Switch `Integer resultCode` to `GsResultCode resultCode` and make changes to submodules also by replacing the null values with corresponding status codes. This makes it easier to debug session state changes as to why it occurs.

karivatj commented on core/src/de/golfgl/gdxgamesvcs/IGameServiceClient.java:
```diff
@@ -278,22 +299,30 @@ boolean fetchLeaderboardEntries(String leaderBoardId, int limit, boolean related
-        GameStateStorage,
+        FetchLeaderBoardEntries,
         GameStateDelete,
         GameStateMultipleFiles,
```
Cleanup and arrange alphabetically

karivatj commented on core/src/de/golfgl/gdxgamesvcs/IGameServiceClient.java:
```diff
@@ -278,22 +299,30 @@ boolean fetchLeaderboardEntries(String leaderBoardId, int limit, boolean related
+        FetchAchievements,
         FetchGameStates,
-        GameStateStorage,
+        FetchLeaderBoardEntries,
```
Cleanup and arrange alphabetically

karivatj commented:
I've made the changes that were requested. I renamed one enum type GsErrorType to GsResultCode and added some success enums there also that can be used as session inactive / active resultcodes.

I will take a look at the sign in stuff discussed earlier a bit more thoroughly when I prepare the PR for the android-gpgs refactor.

karivatj commented on html-gpgs/src/de/golfgl/gdxgamesvcs/GpgsClient.java:
```diff
@@ -267,17 +267,15 @@ private native void refreshDisplayname() /*-{
 
     protected void onDisplayName(String displayName) {
         this.displayName = displayName;
-        if (gsListener != null)
```
Why is this OnSessionActive callback here? Seems weird when we receive a new displayname we trigger a session active event? Is this something that is needed in HTML GPGS stuff? I decided to remove this because I thought it is illogical for this to be here.

MrStahlfelge replied:
Before your change, player display name was guarantueed to be available when game service was connected. Thus, the connection state was set to connected after the display name was received.

karivatj replied:
Okay, I will revert that change. Thanks for the clarification.

karivatj commented on core/src/de/golfgl/gdxgamesvcs/IGameServiceListener.java:
```diff
@@ -27,7 +31,19 @@
-    public void gsShowErrorToUser(GsErrorType et, String msg, Throwable t);
+    public void gsShowErrorToUser(GsResultCode et, String msg, Throwable t);
 
-    public enum GsErrorType {errorLoginFailed, errorUnknown, errorServiceUnreachable, errorLogoutFailed}
```
Rename GsErrorType to GsResultCode. Makes it usable in success cases also. Relevant to for example on session change events.

MrStahlfelge replied:
What different success types can we have?

karivatj replied:
Instead of defining only some generic error types for example `errorLoginFailed `or `errorLogoutFailed` which we use to inform the client when an error occurs, we could define some other status codes as well like `signedIn`, `signedOut`, `connectionPaused`, `connectionResumed `etc and rename the enum type to `GsResultCode`. We could then use it with `OnSessionAcrive` and `OnSessionInactive` methods to inform the client the reason why the session state changed.

Does this idea sound reasonable or is this completely useless thing to consider. What do you think?

MrStahlfelge replied:
Not useless, but at the state GPGS was there was no reliable information what exactly happened, might have changed now.

karivatj commented:
I managed to figure out the sign in process. If the user explicitly signs out it is handled within the GPGS client like previously.

The reason I added the status code to session state changes was to inform the client about the explicit sign out event so the client could persist the choice and make a decision during app startup whether or not to sign in the user. I made a fix to the gpgs client code which handles this behind the scenes so the original use case is not valid anymore (for my case).

It is up to you if you see this as useful. Should the session state change include a reason code? If not, I can remove the commit from the Pull Request in order to get this approved.

Cheers,
Kari

MrStahlfelge commented:
I think we should remove it, I can't see any use case and it breaks backwards compatibility (every game has to change the callbacks).

Ping me when the PR is ready to review.

karivatj commented:
Alrighty, I modified the PR by removing the commit containing the result code stuff.
